File: Documentacion_Plataforma_Inspectores_v2_6/implementation/src/seguridad_vial/normalization.py
from __future__ import annotations

import re
import unicodedata

_SPACE_RE = re.compile(r"\s+")
_DOT_RE = re.compile(r"\s*\.\s*")


def collapse_spaces(value: str) -> str:
    return _SPACE_RE.sub(" ", value.strip())


def normalize_display_name(value: str) -> str:
    value = unicodedata.normalize("NFKC", value or "")
    value = value.replace("’", "'").replace("`", "'")
    value = collapse_spaces(value)
    value = _DOT_RE.sub(".", value)
    return value


def normalize_key(value: str) -> str:
    value = normalize_display_name(value)
    value = unicodedata.normalize("NFD", value)
    value = "".join(ch for ch in value if unicodedata.category(ch) != "Mn")
    value = value.casefold()
    value = re.sub(r"[^a-z0-9]+", " ", value)
    return collapse_spaces(value)


def normalize_sheet_name(value: str) -> str:
    return normalize_key(value)


def normalize_code(value: object) -> str:
    if value is None:
        return ""
    return collapse_spaces(str(value)).upper().replace(" ", "")

File: apps/api/src/modules/bootstrap-initialization/excel/normalization.ts
const SPACE_RE = /\s+/g;
const DOT_RE = /\s*\.\s*/g;

export function collapseSpaces(value: string): string {
  return value.trim().replace(SPACE_RE, ' ');
}

export function normalizeDisplayName(value: string): string {
  let next = (value || '').normalize('NFKC');
  next = next.replace(/[’`]/g, "'");
  next = collapseSpaces(next);
  next = next.replace(DOT_RE, '.');
  return next;
}

export function normalizeKey(value: string): string {
  let next = normalizeDisplayName(value);
  next = next.normalize('NFD').replace(/\p{M}/gu, '');
  next = next.toLowerCase();
  next = next.replace(/[^a-z0-9]+/g, ' ');
  return collapseSpaces(next);
}

export function normalizeSheetName(value: string): string {
  return normalizeKey(value);
}

export function normalizeCode(value: unknown): string {
  if (value === null || value === undefined) {
    return '';
  }
  return collapseSpaces(String(value)).toUpperCase().replace(/ /g, '');
}
